### PR TITLE
Parameters queue

### PR DESCRIPTION
**Pull request by ispielma**
Converts `self.last_out_params` into a queue.  No behavior is changed.

This is part of a larger change that enables M-LOOP to presubmit a number of parameters to the experiment in advance and accept the cost values in sequence as they are provided.  The reason for this is that our experiments really need to be repeating with a specific duty cycle to be stable and M-LOOP needs to optimize within that constraint.  Keeping several shots queued up on the apparatus allows this behavior.  I will be submitting these in bite-sized pieces.

### CODE REVIEW

**Comment by zakv** [comment]
I definitely hear you about wanting to keep your experiment running with a consistent duty cycle. I think there are potentially already a few ways to do this though, so I want to check if those work for you:

First, have you tried the `no_delay` option for M-LOOP? It causes M-LOOP to always suggest a new parameter set immediately. It will suggest parameters from the machine learner if a set is available, otherwise it will use one of the training algorithms (which evaluate very quickly) to suggest a new parameter set right away.

Second, if I recall correctly, blacs has some sort of idle job feature (maybe it's a plugin?) which causes it to run a pre-selected job whenever the job queue empties. That would have the benefit that it would keep your machine warm constantly - not just during an M-LOOP optimization - if desired.

The code in M-LOOP can be hard to follow and has pretty limited test coverage, which means it's very easy to accidentally introduce bugs. So I'm generally hesitant to add new features unless they are pretty minimal/straightforward or very necessary. That said, I certainly want you to be able to use it, so if those alternatives don't work for your use case and it's critical for you that M-LOOP can suggest multiple parameter sets at once, then I'm ok with moving forward with the plans you've mentioned in this PR.

**Comment by ispielma** [comment]
The `no_delay` feature still requires all of the lyse steps before `M-LOOP `to occur as well as the `runmanager` compilation to take place.  This ends up adding ~5 seconds between shots that are not present when we just submit a batch of shots all at once from `runmanager`.

I already have the desired feature working in the lab with `M-LOOP` + `analysislib-mloop` (it makes sure that there are always `N` shots in the `blacs` queue so there is never any delay), and I have done some regression testing to make sure that the old features all _seem_ to work if the presubmit depth is set to `N=1`.  It works well with GaussianProcesses and NeuralNet learners without any change to their code.  I had to rewrite the Random learner (thus identifying the bug above).

Right now I have a branch of `analysislib-mloop` that monkey patches the needed changes into `M-LOOP`, but I am trying to do these pull requests one at a time to eliminate the need for this sort of workaround.

**Comment by zakv** [comment]
Gotcha, sounds good. Happy to move forward with this then, and the changes in this PR in particular look ok so I'll go ahead and merge it.

You may already be aware of this, but just in case this might avoid a headache in the future, FYI the Gaussian process learner in particular can be very slow to suggest new parameters values after many iterations. Depending on the number of parameters, the number of optimization iterations, and the experiment execution time, it may actually reach the point where it can't fill the output parameters queue as quickly as your experiment can empty it - which would then cause idle time. Just something to keep an eye out for, particularly since it wouldn't crop up until later on in an optimization, at which point you may not be watching.
